File: Medium/Java/54. Spiral Matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int top = 0;
        int bot = matrix.length - 1;
        int left = 0;
        int right = matrix[0].length - 1;
        List<Integer> spiral = new ArrayList<>();

        // Find all the elements of the matrix in spiral order
        while(top <= bot && left <= right){
            // Top row
            for(int i = left; i <= right; i++)
                spiral.add(matrix[top][i]);
            top++;

            // Right column
            for(int i = top; i <= bot; i++)
                spiral.add(matrix[i][right]);
            right--;

            // Bottom row
            if(top <= bot){
                for(int i = right; left <= i; i--)
                    spiral.add(matrix[bot][i]);
                bot--;
            }

            // Left column
            if(left <= right){
                for(int i = bot; top <= i; i--)
                    spiral.add(matrix[i][left]);
                left++;
            }
        }

        return spiral;
    }
}

File: Medium/JavaScript/54. Spiral Matrix.js
var spiralOrder = function(matrix) {
    let top = 0;
    let bot = matrix.length - 1;
    let left = 0;
    let right = matrix[0].length - 1;
    let spiral = [];

    // Find all the elements of the matrix in spiral order
    while(top <= bot && left <= right){
        // Top row
        for(let i = left; i <= right; i++)
            spiral.push(matrix[top][i]);
        top++;

        // Right column
        for(let i = top; i <= bot; i++)
            spiral.push(matrix[i][right]);
        right--;

        // Bottom row
        if(top <= bot){
            for(let i = right; left <= i; i--)
                spiral.push(matrix[bot][i]);
            bot--;
        }

        // Left column
        if(left <= right){
            for(let i = bot; top <= i; i--)
                spiral.push(matrix[i][left]);
            left++;
        }
    }

    return spiral;
};
